### ISSUE TITLE
[Question] - How to use -detailed-exitcode with terragrunt plan

### ISSUE THREAD

**Issue by SurdTechnologies** [gruntwork-io/terragrunt#1470]
https://www.terraform.io/docs/commands/plan.html has a 
-detailed-exitcode - Return a detailed exit code when the command exits. When provided, this argument changes the exit codes and their meanings to provide more granular information about what the resulting plan contains:

0 = Succeeded with empty diff (no changes)
1 = Error
2 = Succeeded with non-empty diff (changes present)

When I try to use terragrunt plan -detailed-exitcode I am getting error and my script is not proceeding.

**Comment by SurdTechnologies** [comment]
`Running the command `terragrunt plan -detailed-exitcode` gives

```
Running command: terraform plan -detailed-exitcode
Hit multiple errors:
exit status 2
```

**Comment by SurdTechnologies** [comment]
It seems the error is there only when plan detects there is a change. If there is no change ( exit status 0 ), the script doesn't stop/exit.
The workaround so far is to run terragrunt plan > out.txt ( without the -detailed-exitcode ) and then parse out.txt to see if any changes are there and take according action.

**Comment by yorinasub17** [comment]
Unfortunately, we don't currently support this in terragrunt because we don't do any special parsing of the args for terraform, and thus are not intelligently handling the non-zero exit code for non-empty diff. The other issue is that terragrunt does not exit with the error code from terraform (this is due to the ambiguities introduced with using `xxx-all` commands).

To support this, we probably need to enhance the CLI parsing to handle `-detailed-exitcode`, taking care to throw a warning or error when using it with `xxx-all` commands.

**Comment by vainkop** [comment]
Looks like it's not easy to detect infra drift & send the diff to an issue/slack until it matches terraform.

**Comment by GuptaNavdeep1983** [comment]
> It seems the error is there only when plan detects there is a change. If there is no change ( exit status 0 ), the script doesn't stop/exit. The workaround so far is to run terragrunt plan > out.txt ( without the -detailed-exitcode ) and then parse out.txt to see if any changes are there and take according action.

May be its possible to use terragrunt plan -out=<name>.tfstate followed by terragrunt show -json to get the plan changes as a json and use the json to decide further course of action for the new plan.

**Comment by ashlineldridge** [comment]
Terragrunt support for `-detailed-exitcode` would make Terragrunt a lot more usable in CI/CD pipelines like GitHub Actions where typically you don't want to generate the apply step if the plan step produces an empty plan.

**Comment by Mikedu1988** [comment]
is there a plan for this feature?

**Comment by mwos-sl** [comment]
Oh that would enable implementing this: https://github.com/runatlantis/atlantis/issues/4742

**Comment by k2m30** [comment]
any updates?

**Comment by yhakbar** [comment]
Hey folks,

This is worth prioritizing, as it seems Terragrunt is handling this behavior in a buggy fashion:

```hcl
# main.tf
// Create a local file
resource "local_file" "example" {
  content  = "This is an example file"
  filename = "${path.module}/example.txt"
}

# terragrunt.hcl
// Empty file
```

```bash
$ terragrunt plan -detailed-exitcode
...
08:13:19.754 ERROR  tofu invocation failed in .
08:13:19.755 ERROR  error occurred:

* Failed to execute "tofu plan -detailed-exitcode" in .

  exit status 2
```

We'll look to tackle this soon.

Thanks for bringing this up, and your patience on waiting for it to be resolved!

**Comment by levkohimins** [comment]
Resolved in [v0.69.1](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.69.1) release.
